core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlPrepareHandlerBase.java

```diff
@@ -171,25 +171,25 @@ public void onThrottleReady(boolean wasDelayed) {
     sendRequest(null, 0);
   }
 
-  private ScheduledFuture<?> scheduleTimeout(Duration timeout) {
-    if (timeout.toNanos() > 0) {
-      return scheduler.schedule(
-          () -> {
-            setFinalError(new DriverTimeoutException("Query timed out after " + timeout));
+  private Timeout scheduleTimeout(Duration timeoutDuration) {
+    if (timeoutDuration.toNanos() > 0) {
+      return this.timer.newTimeout(
```

Comment:
can we extract common logic for scheduling timeout for CqlRequestHandlerBase and CqlPrepareHandlerBase?
We could create a separate class that is responsible only for timeouts and reuse it for both classes

Reply:
Code reuse in request handlers is a dilemma: they have a lot of things in common, but also a lot of differences. For example both do timeouts but CqlPrepareHandlerBase doesn't have speculative executions, CqlRequestHandlerBase doesn't have to deal with repreparing on other nodes, etc.
If we start to extract some common logic I'm worried it's going to become an ugly abstraction with conditionals everywhere. That also raises the question of exposing that logic for 3rd-party handler implementations, which could make its design even more complicated.

So granted there is a bit of code duplication with the current design, but on the other hand each handler is its own self-contained implementation, that is easy to understand and modify.

Reply:
Thanks for a detailed explanation